### PR TITLE
Removing CloudPickle dependency for mlflow pytorch log/save model

### PR DESCRIPTION
**Pull request by shrinath-suresh**
## What changes are proposed in this pull request?

https://pytorch.org/tutorials/beginner/saving_loading_models.html 

Pytorch now supports `pickle` module for serialization. Removing cloud pickle dependencies. 

## How is this patch tested?

Tested save, log and load model with pickle as serialization module. Models are getting saved and loaded. pickle_module_info.txt is showing pickle as well. 

### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?
Components 
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [x] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: Local serving, model deployment tools, spark UDFs
- [ ] `area/server-infra`: MLflow server, JavaScript dev server
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface 
- [ ] `area/uiux`: Front-end, user experience, JavaScript, plotting
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language 
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [x] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/pytorch/__init__.py [hunk 1]
```diff
@@ -52,12 +52,6 @@ def get_default_conda_env():
             "pytorch={}".format(torch.__version__),
             "torchvision={}".format(torchvision.__version__),
         ],
-        additional_pip_deps=[
-            # We include CloudPickle in the default environment because
-            # it's required by the default pickle module used by `save_model()`
-            # and `log_model()`: `mlflow.pytorch.pickle_module`.
-            "cloudpickle=={}".format(cloudpickle.__version__)
```
Comment (shrinath-suresh) [inline mlflow/pytorch/__init__.py:59]: will add pickle module here

**Comment by dbczumar** [comment]
@shrinath-suresh @smurching With this change, can we still capture PyTorch model classes defined in the `__main__` scope or a notebook environment?

**Comment by shrinath-suresh** [comment]
> @shrinath-suresh @smurching With this change, can we still capture PyTorch model classes defined in the `__main__` scope or a notebook environment?

Thanks for the comment. Looks like there are specific tests written for it - https://github.com/mlflow/mlflow/blob/326c5ecc77b16bc8b08ecc504d62dc5183be1cfa/tests/pytorch/test_pytorch_model_export.py#L506 . When i change the pickle_module from `cloudpickle` to `cloudpickle` , save_model is throwing exception. So the answer is `No`

**Comment by shrinath-suresh** [comment]
> > @shrinath-suresh @smurching With this change, can we still capture PyTorch model classes defined in the `__main__` scope or a notebook environment?
> 
> Thanks for the comment. Looks like there are specific tests written for it -
> 
> https://github.com/mlflow/mlflow/blob/326c5ecc77b16bc8b08ecc504d62dc5183be1cfa/tests/pytorch/test_pytorch_model_export.py#L506
> 
> . When i change the pickle_module from `cloudpickle` to `pickle` , save_model is throwing exception. So the answer is `No`

@dbczumar @smurching Can you please suggest if we need to live with `cloudpickle` dependency (even for newer version of pytorch models) due to the above mentioned issue.

**Review by smurching — COMMENTED** [review]
@shrinath-suresh it seems we may need to stick with cloudpickle for now for the reasons @dbczumar mentioned. 

Our existing pytorch example ([link](https://github.com/mlflow/mlflow/blob/1f637f888445df2e6a1e89d3c164d467de62eef0/examples/pytorch/mnist_tensorboard_artifact.py#L226)) actually calls `mlflow.pytorch.log_model` -> `torch.save` using `pickle` rather than `cloudpickle` (i.e. it uses `pickle`, as this PR attempts to do). 

However, this causes loading back the model afterwards with `mlflow.pytorch.load_model` to break:


```
$ python examples/pytorch/mnist_tensorboard_artifact.py --epochs=0
...
Logging the trained model as a run artifact...

The model is logged at:
./mlruns/0/6e67a69c815145ae9f438ce6ad4fc18f/artifacts/pytorch-model

$  python
>>> import mlflow.pytorch
>>> model = mlflow.pytorch.load_model("./mlruns/0/6e67a69c815145ae9f438ce6ad4fc18f/artifacts/pytorch-model")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/sid.murching/code/mlflow/mlflow/pytorch/__init__.py", line 451, in load_model
    return _load_model(path=torch_model_artifacts_path, **kwargs)
  File "/Users/sid.murching/code/mlflow/mlflow/pytorch/__init__.py", line 393, in _load_model
    return torch.load(model_path, **kwargs)
  File "/Users/sid.murching/miniconda3/envs/mlflow-python3/lib/python3.6/site-packages/torch/serialization.py", line 592, in load
    return _load(opened_zipfile, map_location, pickle_module, **pickle_load_args)
  File "/Users/sid.murching/miniconda3/envs/mlflow-python3/lib/python3.6/site-packages/torch/serialization.py", line 852, in _load
    result = unpickler.load()
AttributeError: Can't get attribute 'Net' on <module '__main__' (built-in)>

```

**Comment by smurching** [comment]
It does seem users run into problems with cloudpickle quite often:
* https://github.com/mlflow/mlflow/issues/3199
* https://github.com/mlflow/mlflow/issues/3258 (underlying issue seems related to Docker, but user was confused by the fact that we call `torch.save` instead of saving the model's state_dict, which is the [recommended approach](https://pytorch.org/tutorials/beginner/saving_loading_models.html#save-load-state-dict-recommended))

I wonder if we should try to support the `state_dict`-based persistence API & unify support for that & torchscript e.g something like:

```
import torch
import mlflow.pytorch
class MyCustomModelClass(torch.nn.Module):

# Train model
my_model = MyCustomModelClass(...)
model.train()
...
# Save model with MLflow. Saves both the model and its state dict. Generates a warning if unable to save the model via 
# torch.save, throws if saving the state_dict fails.
# Users can pass optional save_torchscript = True to additionally save the model in torchscript
mlflow.pytorch.log_model(my_model, "my-custom-path", save_torchscript=True)

# Load model back using the new recommended approach of loading via state dict.
# More robust, doesn't require having `MyCustomModelClass` in a file at some specific 
# path at model-load time, but puts the burden on the user to ensure `MyCustomModelClass` definition is available
mlflow.pytorch.load_model("my-custom-path", model_class=MyCustomModelClass)

# Or load pickled torch.save'd model back:
mlflow.pyotrch.load_model("my-custom-path")

```

**Comment by shrinath-suresh** [comment]
@smurching For main scoped models using pickle i have added a new test case to validate the attribute error. Can you please review the code and lemme know your comments?

**Comment by smurching** [comment]
@shrinath-suresh apologies for the delay & thanks for adding the test! MLflow's APIs are stable, so we unfortunately can't go with the solution currently in this PR as it will break existing user code, i.e. cause loading of main-scoped Pytorch models to fail.

To write it out, this is my current understanding of the use cases supported by MLflow, along with the corresponding approach recommended by Pytorch:

* Loading models back for retraining via `mlflow.pytorch.load_model`. For this use case, Pytorch recommends saving & loading models using their `state_dict` and class. This works well for users, as usually they’re loading models back in the context of existing training code/in the same Python environment in which the model was originally trained, so the class definition is readily available
* Loading models back for inference, e.g. for batch inference via `mlflow.pyfunc.spark_udf("/path/to/my/logged/pytorch/model"), or RESTful inference via `mlflow models serve /path/to/my/logged/pytorch/model`. Pytorch recommends using torchscript for inference, as it's both more performant and hermetic. The latter property is particularly important, as model inference often happens in different environments from model training, so having a fully self-contained model is critical.

Thus I’m wondering if the right approach is to try to use the right pytorch APIs/persistence format for the appropriate scenario. That is, if MLflow should try to both save the model’s state dict and its torchscript representation when `mlflow.pytorch.log_model` is called, so that `mlflow.pytorch.load_model` can attempt to load back the model using the state dict & a user-supplied class, and inference APIs like `mlflow models serve` can depend on the model’s self-contained torchscript representation. I think we'd ultimately need to still support saving/loading eager models via `mlflow.pytorch.log_model` and `mlflow.pytorch.load_model` to avoid breaking existing user code, but maybe we can extend these APIs to support state_dict and torchscript, as in the strawman proposal in https://github.com/mlflow/mlflow/pull/3408/#issuecomment-693013610. Note that I'm not wedded to the API proposed in my comment - for example, maybe an API like `mlflow.pytorch.load_state_dict(model_uri)` makes more sense than `mlflow.pytorch.load_model(model_uri, model_class=None)`

Let me know if that makes sense / what you think!

cc also @harupy

**Pull request closed without merging**